c-bindings-gen/src/types.rs

```diff
@@ -0,0 +1,1102 @@
+use std::collections::HashMap;
+
+use proc_macro2::{TokenTree, Span};
+
+// We have a few main types -
+// * Unit Enums, which are "mirrored" by creating a #[repr(C)] version and exposing it.
+// * Traits, which map to a void* and a jump table,
+// * Opaque Structs, which are exposed as-is either directly or via a typedef if they are generic
+//   (using our own concrete traits, which are enumerated below).
+//   All pub methods are exposed and pub fields are exposed via geters/setters.
+
+// The following types are used purely to build our known types maps - they break down all the
+// types we need to resolve to include the given object, and no more.
+
+pub fn first_seg_self<'a>(t: &'a syn::Type) -> Option<impl Iterator<Item=&syn::PathSegment> + 'a> {
+	match t {
+		syn::Type::Path(p) => {
+			if p.qself.is_some() || p.path.leading_colon.is_some() {
+				return None;
+			}
+			let mut segs = p.path.segments.iter();
+			let ty = segs.next().unwrap();
+			if !ty.arguments.is_empty() { return None; }
+			if format!("{}", ty.ident) == "Self" {
+				Some(segs)
+			} else { None }
+		},
+		_ => None,
+	}
+}
+
+pub fn get_single_remaining_path_seg<'a, I: Iterator<Item=&'a syn::PathSegment>>(segs: &mut I) -> Option<&'a syn::Ident> {
+	if let Some(ty) = segs.next() {
+		if !ty.arguments.is_empty() { unimplemented!(); }
+		if segs.next().is_some() { return None; }
+		Some(&ty.ident)
+	} else { None }
+}
+
+pub fn assert_single_path_seg<'a>(p: &'a syn::Path) -> &'a syn::Ident {
+	if p.leading_colon.is_some() { unimplemented!(); }
+	get_single_remaining_path_seg(&mut p.segments.iter()).unwrap()
+}
+
+pub fn single_ident_generic_path_to_ident(p: &syn::Path) -> Option<&syn::Ident> {
+	if p.segments.len() == 1 {
+		Some(&p.segments.iter().next().unwrap().ident)
+	} else { None }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum ExportStatus {
+	Export,
+	NoExport,
+	TestOnly,
+	Rename(String),
+}
+pub fn export_status(attrs: &[syn::Attribute]) -> ExportStatus {
+	for attr in attrs.iter() {
+		let tokens_clone = attr.tokens.clone();
+		let mut token_iter = tokens_clone.into_iter();
+		if let Some(token) = token_iter.next() {
+			match token {
+				TokenTree::Punct(c) if c.as_char() == '=' => {
+					// Really not sure where syn gets '=' from here -
+					// it somehow represents '///' or '//!'
+				},
+				TokenTree::Group(g) => {
+					if format!("{}", single_ident_generic_path_to_ident(&attr.path).unwrap()) == "cfg" {
+						if let TokenTree::Ident(i) = g.stream().into_iter().next().unwrap() {
+							if i == "test" || i == "feature" {
+								// If its cfg(feature(...)) we assume its test-only
+								return ExportStatus::TestOnly;
+							}
+						}
+					}
+					continue; // eg #[derive()]
+				},
+				_ => unimplemented!(),
+			}
+		} else { continue; }
+		match token_iter.next().unwrap() {
+			TokenTree::Literal(lit) => {
+				let line = format!("{}", lit);
+				if line.contains("(C-not exported)") {
```

Comment:
Hmm. We could change the line, but I actually kinda like putting it in the docs - it lets us link people to the rustdoc output even if they're using the library via C.

Reply:
I'm not sure I follow the part about linking docs.

What I meant was we should define attributes that we use to annotate rust-lightning. Then this code checks for that attribute rather than trying to parse Rust docs.

Reply:
Right, the use of docs serves a dual purpose - both to allow us to make decisions at codegen time, and to actually document the fact that something isn't exported. Currently I've been sending people to the docs.rs docs for C bindings as well, though we could try to change that.